File: 算法笔记/NQSDemo/src/main/java/jeason/b树形结构/a二叉搜索树/Printer工具类/BinaryTrees.java

package jeason.b树形结构.a二叉搜索树.Printer工具类;

import jeason.b树形结构.a二叉搜索树.Printer工具类.base.BinaryTreeInfoProtocol;
import jeason.b树形结构.a二叉搜索树.Printer工具类.base.Printer;

public final class BinaryTrees {

	private BinaryTrees() {
	}

	public static void print(BinaryTreeInfoProtocol tree) {
		print(tree, null);
	}

	public static void println(BinaryTreeInfoProtocol tree) {
		println(tree, null);
	}

	public static void print(BinaryTreeInfoProtocol tree, PrintStyle style) {
		if (tree == null || tree.root() == null) return;
		printer(tree, style).print();
	}

	public static void println(BinaryTreeInfoProtocol tree, PrintStyle style) {
		if (tree == null || tree.root() == null) return;
		printer(tree, style).println();
	}

	public static String printString(BinaryTreeInfoProtocol tree) {
		return printString(tree, null);
	}

	public static String printString(BinaryTreeInfoProtocol tree, PrintStyle style) {
		if (tree == null || tree.root() == null) return null;
		return printer(tree, style).printString();
	}

	private static Printer printer(BinaryTreeInfoProtocol tree, PrintStyle style) {
		if (style == PrintStyle.INORDER) return new InorderPrinter(tree);
		return new LevelOrderPrinter(tree);
	}

	public enum PrintStyle {
		LEVEL_ORDER, INORDER
	}
}

File: PythonAlgorithm/Study/b树形结构/a二叉搜索树/BinaryTrees.py

from .Printer工具类.BinaryTree_AbstractInfo import BinaryTree_AbstractInfo
from .Printer工具类.Printer_Inorder import Printer_Inorder
from .Printer工具类.Printer_Levelorder import Printer_Levelorder
from enum import Enum



class PrintStyle(Enum):
    LEVEL_ORDER = 0
    INORDER = 1

class BinaryTrees(object):
    def __init__(self):
        pass
            
    @staticmethod
    def print(tree, style):
        """单纯打印"""
        if tree is None or (tree.root() is None):
            return
        BinaryTrees.generate_printer(tree, style).print()
    
    @staticmethod
    def println(tree, style):
        """打印后换行"""
        if tree is None or (tree.root() is None):
            return
        BinaryTrees.generate_printer(tree, style).println()
        
    @staticmethod
    def printString(tree, style):
        """生成整个树状图的字符串"""
        if tree is None or tree.root() is None:
            return None
        return BinaryTrees.generate_printer(tree, style).printString()



    @staticmethod
    def generate_printer(tree: BinaryTree_AbstractInfo, style: PrintStyle):
        if style == PrintStyle.INORDER:
            return Printer_Inorder(tree)
        return Printer_Levelorder(tree)